File: code/21-2-chronal-conversion/solve.py
#!/usr/bin/env pypy3 -O


def solve(t):
    prog = t.splitlines()
    rip = int(prog[0].split()[-1])
    prog = [(op, *map(int, xs)) for s in prog[1:] for op, *xs in [s.split()]]

    def addr(a, b, c, mem):
        mem[c] = mem[a] + mem[b]
    def addi(a, b, c, mem):
        mem[c] = mem[a] + b
    def bani(a, b, c, mem):
        mem[c] = mem[a] & b
    def bori(a, b, c, mem):
        mem[c] = mem[a] | b
    def eqri(a, b, c, mem):
        mem[c] = 1 if mem[a] == b else 0
    def eqrr(a, b, c, mem):
        mem[c] = 1 if mem[a] == mem[b] else 0
    def gtir(a, b, c, mem):
        mem[c] = 1 if a > mem[b] else 0
    def gtrr(a, b, c, mem):
        mem[c] = 1 if mem[a] > mem[b] else 0
    def muli(a, b, c, mem):
        mem[c] = mem[a] * b
    def seti(a, b, c, mem):
        mem[c] = a
    def setr(a, b, c, mem):
        mem[c] = mem[a]

    instrs = dict(
        addi=addi, addr=addr,
        bani=bani, bori=bori,
        eqri=eqri, eqrr=eqrr,
        gtir=gtir, gtrr=gtrr,
        muli=muli,
        seti=seti, setr=setr
    )

    mem = [0] * 6
    ip = 0
    res = None
    visited = set()

    while 0 <= ip < len(prog):
        # part 2
        if ip == 29:
            if mem[5] in visited:
                break
            res = mem[5]
            visited.add(res)

        mem[rip] = ip
        op, a, b, c = prog[ip]
        instrs[op](a, b, c, mem=mem)
        ip = mem[rip] + 1

    return res


if __name__ == '__main__':
    import fileinput
    with fileinput.input() as f:
        text = ''.join(f).strip('\n')

    print(solve(text))

File: code/21-2-chronal-conversion/solve.cpp
// c++ -O2 solve.cpp -o solve
#include <stdint.h>
#include <stdio.h>
#include <unordered_set>


using namespace std;


typedef enum opcodes {
    addi,
    addr,
    bani,
    bori,
    eqri,
    eqrr,
    gtir,
    gtrr,
    muli,
    seti,
    setr,
} opcodes;


typedef uint8_t u8;
typedef uint32_t u32;


static const u8 rip = 2;

static const u32
prog[][4] = {
    {seti, 123, 0, 5},
    {bani, 5, 456, 5},
    {eqri, 5, 72, 5},
    {addr, 5, 2, 2},
    {seti, 0, 0, 2},
    {seti, 0, 9, 5},
    {bori, 5, 65536, 3},
    {seti, 7586220, 4, 5},
    {bani, 3, 255, 1},
    {addr, 5, 1, 5},
    {bani, 5, 16777215, 5},
    {muli, 5, 65899, 5},
    {bani, 5, 16777215, 5},
    {gtir, 256, 3, 1},
    {addr, 1, 2, 2},
    {addi, 2, 1, 2},
    {seti, 27, 9, 2},
    {seti, 0, 9, 1},
    {addi, 1, 1, 4},
    {muli, 4, 256, 4},
    {gtrr, 4, 3, 4},
    {addr, 4, 2, 2},
    {addi, 2, 1, 2},
    {seti, 25, 4, 2},
    {addi, 1, 1, 1},
    {seti, 17, 2, 2},
    {setr, 1, 6, 3},
    {seti, 7, 8, 2},
    {eqrr, 5, 0, 1},
    {addr, 1, 2, 2},
    {seti, 5, 0, 2},
};


static inline void
dispatch(const u32* instr, u32* mem) {
    switch (instr[0]) {
        case addi:
            mem[instr[3]] = mem[instr[1]] + instr[2];
            break;
        case addr:
            mem[instr[3]] = mem[instr[1]] + mem[instr[2]];
            break;
        case bani:
            mem[instr[3]] = mem[instr[1]] & instr[2];
            break;
        case bori:
            mem[instr[3]] = mem[instr[1]] | instr[2];
            break;
        case eqri:
            mem[instr[3]] = mem[instr[1]] == instr[2];
            break;
        case eqrr:
            mem[instr[3]] = mem[instr[1]] == mem[instr[2]];
            break;
        case gtir:
            mem[instr[3]] = instr[1] > mem[instr[2]];
            break;
        case gtrr:
            mem[instr[3]] = mem[instr[1]] > mem[instr[2]];
            break;
        case muli:
            mem[instr[3]] = mem[instr[1]] * instr[2];
            break;
        case seti:
            mem[instr[3]] = instr[1];
            break;
        case setr:
            mem[instr[3]] = mem[instr[1]];
            break;
    }
}


int main(int argc, char const *argv[]) {
    const int progn = sizeof(prog) / sizeof(prog[0]);
    u32 mem[6] = {};
    unordered_set<u32> visited;
    u32 res = 0;

    for (int ip = 0; ip >= 0 && ip < progn; ip++) {
        if (ip == 29) {
            if (visited.find(mem[5]) != visited.end()) {
                break;
            }
            res = mem[5];
            visited.insert(res);
        }

        mem[rip] = ip;
        dispatch(prog[ip], mem);
        ip = mem[rip];
    }

    printf("%d\n", res);

    return 0;
}
